patches/@endo+ses-ava+0.2.27.patch

```diff
@@ -0,0 +1,250 @@
+@@ -1 +0,0 @@
+-import './src/types.js';
+diff --git a/node_modules/@endo/ses-ava/src/ses-ava-test.js b/node_modules/@endo/ses-ava/src/ses-ava-test.js
+index 36038ad..00f4114 100644
+--- a/node_modules/@endo/ses-ava/src/ses-ava-test.js
++++ b/node_modules/@endo/ses-ava/src/ses-ava-test.js
+@@ -1,18 +1,8 @@
+ // @ts-check
+ import 'ses';
+-import './types.js';
+ 
+ const { apply } = Reflect;
+ 
+-/**
+- * Just forwards to global `console.error`.
+- *
+- * @type {Logger}
+- */
+-const defaultLogger = (...args) => {
+-  console.error(...args);
+-};
+-
+ /**
+  * Determine if the argument is a Promise.
+  * (Approximately copied from promiseKit.js)
+@@ -24,21 +14,46 @@ const isPromise = maybePromise =>
+   Promise.resolve(maybePromise) === maybePromise;
+ 
+ /**
+- * @type {LogCallError}
++ * @typedef {(...args: unknown[]) => void} Logger
++ */
++
++/**
++ * Calls `thunk()` passing back approximately its outcome, but first
++ * logging any erroneous outcome to the `logError`, which defaults to
++ * `console.error`.
++ *
++ * If thunk returns a non-promise, silently return it.
++ * If thunk throws, log what was thrown and then rethrow it.
++ * If thunk returns a promise, immediately return a new unresolved promise.
++ * If the first promise fulfills, silently fulfill the returned promise even if
++ * the fulfillment was an error.
++ * If the first promise rejects, log the rejection reason and then reject the
++ * returned promise with the same reason.
++ * The delayed rejection of the returned promise is an observable difference
++ * from directly calling `thunk()` but will be equivalent enough for most
++ * purposes.
++ *
++ * TODO This function is useful independent of Ava, so consider moving it
++ * somewhere and exporting it for general reuse.
++ *
++ * @param {(...unknown) => unknown} func
++ * @param {unknown[]} args
++ * @param {string} name
++ * @param {Logger=} logError
+  */
+-const logErrorFirst = (func, args, name, logger = defaultLogger) => {
++const logErrorFirst = (func, args, name, logError = console.error) => {
+   let result;
+   try {
+     result = apply(func, undefined, args);
+   } catch (err) {
+-    logger(`THROWN from ${name}:`, err);
++    logError(`THROWN from ${name}:`, err);
+     throw err;
+   }
+   if (isPromise(result)) {
+     return result.then(
+       v => v,
+       reason => {
+-        logger(`REJECTED from ${name}:`, reason);
++        logError(`REJECTED from ${name}:`, reason);
+         return result;
+       },
+     );
+@@ -47,7 +62,7 @@ const logErrorFirst = (func, args, name, logger = defaultLogger) => {
+   }
+ };
+ 
+-const testerMethodsWhitelist = [
++const overrideList = [
+   'after',
+   'afterEach',
+   'before',
+@@ -56,32 +71,35 @@ const testerMethodsWhitelist = [
+   'failing',
+   'serial',
+   'only',
+-  'skip',
+ ];
+ 
+ /**
+- * @param {TesterFunc} testerFunc
+- * @param {Logger} [logger]
+- * @returns {TesterFunc} Not yet frozen!
++ * @template {Function} T
++ * @param {T} testerFunc
```

Comment:
I don't understand replacing the specific type `TesterFunc` with a generic type parameter. The returned `augmented` func assumes that the `testerFunc`'s second parameter is an `ImplFunc`. The `augmented` function itself is supposed to be a compat replacement for `testerFunc` only under this assumption, since it has its own `implFunc` parameter, which it replaces with `testFuncWrapper`.

If `testerFunc` is anything else, then `wrapTester(testerFunc, ...)` would be wrong. The method names of the `overrideList`/`testerMethodsWhitelist` are supposed to enumerate only methods of `test` which conform to these assumptions about `TesterFunc`.

Reply:
I went with any `Function` because it works. I don't see the value in the additional constraints. If this is a blocker let me know.

Reply:
It is not a blocker. I recognize that you understand our types much better than I. So when I'm puzzled by your use of types, I'd at least like to learn from you. 

Here, I am genuinely confused about why you think this change is an improvement. The code being described by these types relies on the assumption that the second parameter of the testerFunc is an ImplFunc, in order to replace this ImplFunc by wrapping. If we didn't have a type language to describe it, I'd feel compelled to write prose describing that. But we do have a type language that is good at saying things like this. Why not use it?

I suppose the blocker is that if we do not explain this constraint using the type language, then we do need explicit prose describing it well. Why wouldn't we?

Reply:
> confused about why you think this change is an improvement

I don't necessarily think every line is an improvement. I think the overall PR is an improvement.

Is it a regression? It's a loss of type information but it's now _less wrong_ because not every tester function takes an impl. E.g. `test.todo` https://github.com/endojs/endo/issues/645

> we do have a type language that is good at saying things like this. Why not use it?

Because of the cost/benefit of making it accurate and not misleading. With this PR the costs of not doing it have been going up so I've gone ahead and re-added TesterFunc and made `implFunc` optional.

Reply:
Aha! I think I understand our misunderstanding. You say 
> not every tester function takes an impl. E.g. `test.todo`

But I omitted `todo` from the whitelist, and I see you still omit it from the `overrideList`. Then it occurred to me: I am using `TesterFunc` only to describe the methods on this list. You (and possibly TS itself?) are interpreting it as a description of all methods on `test`, which it clearly is not. If `todo` isn't described by `TesterFunc` then it should not be on this list anyway.

Making `implFunc` optional seems like a solution only because of our misunderstanding.

How should this be expressed in the type system? I thought I had done that successfully by enumerating in `TesterProperties` exactly the methods I put into the list. `todo` is absent both from the list and from the `TesterProperties` type. 

(Having the `TesterProperties` type track the list is a redundancy that creates a refactoring hazards, which I would also like to avoid if possible. But let's take that question separately.)

Reply:
Sounds like a good plan. =)